providers/linode/provider.go

```diff
@@ -0,0 +1,263 @@
+									 curl \
+									 gnupg
+install -m 0755 -d /etc/apt/keyrings
+curl -fsSL https://download.docker.com/linux/ubuntu/gpg | gpg --dearmor -o /etc/apt/keyrings/docker.gpg
+chmod a+r /etc/apt/keyrings/docker.gpg
+
+# Add Docker sources
+echo \
+  "deb [arch=$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.gpg] https://download.docker.com/linux/ubuntu \
+  $(. /etc/os-release && echo "$VERSION_CODENAME") stable" | \
+  tee /etc/apt/sources.list.d/docker.list > /dev/null
+
+apt-get update && apt-get install -y docker-ce docker-compose-plugin
+
+systemctl enable --now docker
+
+cat > /root/docker-compose.yml <<'EOS'
+version: '3'
+services:
+  woodpecker-agent:
+    image: {{ .Image }}
+    restart: always
+    volumes:
+      - /var/run/docker.sock:/var/run/docker.sock
+    environment:
+      {{- range $key, $value := .Environment }}
+        - {{ $key }}={{ $value }}
+      {{- end }}
+EOS
+
+cd /root && docker compose up -d`
+
+// editorconfig-checker-enable
+type Provider struct {
+	region        string
+	name          string
+	instanceType  string
+	image         string
+	config        *config.Config
+	sshKey        string
+	rootPass      string
+	stackscriptID int
+	userData      *template.Template
+	tags          []string
+	client        *linodego.Client
+}
+
+func New(c *cli.Context, config *config.Config) (engine.Provider, error) {
+	d := &Provider{
+		name:          "linode",
+		region:        c.String("linode-region"),
+		instanceType:  c.String("linode-instance-type"),
+		image:         c.String("linode-image"),
+		sshKey:        c.String("linode-ssh-key"),
+		rootPass:      c.String("linode-root-pass"),
+		stackscriptID: c.Int("linode-stackscript-id"),
+		config:        config,
+	}
+
+	d.client = newClient(c.String("linode-api-token"))
+
+	ctx := context.Background()
+	err := d.setupKeypair(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("%s: setupKeypair: %w", d.name, err)
+	}
+
+	userDataStr := engine.CloudInitUserDataUbuntuDefault
+	// TODO: remove once linode user-data feature is out of beta
+	if d.stackscriptID != -1 {
+		userDataStr = stackscriptUserDataDefault
+	}
+	if _userDataStr := c.String("linode-user-data"); _userDataStr != "" {
+		userDataStr = _userDataStr
+	}
+	userDataTmpl, err := template.New("user-data").Parse(userDataStr)
+	if err != nil {
+		return nil, fmt.Errorf("%s: template.New.Parse %w", d.name, err)
+	}
+	d.userData = userDataTmpl
+
+	d.tags = c.StringSlice("linode-tags")
+
+	return d, nil
+}
+
+func (d *Provider) DeployAgent(ctx context.Context, agent *woodpecker.Agent) error {
+	userdataString, err := engine.RenderUserDataTemplate(d.config, agent, d.userData)
+	if err != nil {
+		return fmt.Errorf("%s: RenderUserDataTemplate: %w", d.name, err)
+	}
+	userData := b64.StdEncoding.EncodeToString([]byte(userdataString))
+
+	userdataMap := make(map[string]string)
+
+	var metadata *linodego.InstanceMetadataOptions
+
+	// TODO: remove once linode user-data is out of beta
```

Comment:
The user data service was out of beta January 2nd 2024.
https://www.linode.com/blog/compute/customize-deployments-with-akamais-metadata-service/

Reply:
It is still not available in all data centres.  For me the closest is ap-southeast in Sydney.  This region does not have metadata service